### ISSUE TITLE
Allow for comparing values of different types

### ISSUE THREAD

**Issue by dalexeev** [godotengine/godot-proposals#4571]
### Describe the project you are working on

A game

### Describe the problem or limitation you are having in your project

**Note:** This proposal is NOT about comparing values of different types like in JavaScript and PHP, I'm NOT asking for `1 == "1"` to be `true`, I'm asking for `false` instead of an error.

* godotengine/godot#26375

It is not possible to compare values of different types using the `==` operator.

![](https://user-images.githubusercontent.com/47700418/169646211-1eaba968-5d9e-4dc6-b525-e394d5d32a9d.png)

I expect that if the types are different, the `==` operator should return `false` instead of an error. IMO an error with different types of operands is too strict. In most languages, the comparison operator does not raise errors (although the behavior varies). For those cases where it is known in advance that the operands are of different types, we can display warnings/notices (see below).

There are currently two exceptions where you can compare values of different types:

1. Comparison of any value (except `null`) with `null` gives false. Does not conflict with this proposal.
2. Comparison of `int` with `float` gives `true` if the values are numerically equal. Controversial, since floats are recommended to be compared using `is_equal_approx`.

Also, `NAN == NAN` is `false` and `NAN != NAN` is `true`. But that doesn't matter in this proposal.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

```gdscript
1 == "str"    # false (now error)
1 == "1"      # false (now error)
1 == 2        # false
1 == 1        # true

1 == null     # false
0 == null     # false
false == null # false
null == null  # true

1 == 1.0      # false (now true)
1 == 2.0      # false
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Additionally we can display the "Condition is always true/false" warning/notice if it is known from the static types of the operands:

```gdscript
const A = 1
var a := 1
var x = 1

if A == 1: pass     # Warning "Condition is always true"
if A == 2: pass     # Warning "Condition is always false"
if A == "str": pass # Warning "Condition is always false"

if a == 1: pass
if a == 2: pass
if a == "str": pass # Warning "Condition is always false"

if x == 1: pass
if x == 2: pass
if x == "str": pass
```

This will help avoid unintentional errors for those using static typing.

`a != b` must be the opposite of `a == b`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

```gdscript
class_name Util

static func equal(a, b) -> bool:
    return typeof(a) == typeof(b) and a == b
```

### Is there a reason why this should be core and not an add-on in the asset library?

This is about GDScript.

**Comment by Calinou** [comment]
> **Note:** This proposal is NOT about comparing values of different types like in JavaScript and PHP, I'm NOT asking for 1 == "1" to be true, I'm asking for false instead of an error.

Even if you return `false` in those scenarios, I still think this leaves too much room for ambiguity and accidental errors to pop up.

**Comment by ywmaa** [comment]
I don't see this as a good feature, I think it will make it worse for me, I can compare a string with a number not on purpose which should give me an error, I am also not sure when will ever anyone need to compare values of different type (except for null with false for example, as null is similar to false, though it is entirely something different)

a warning won't help me a lot because when I code in godot a warning should not affect the flow of the code, though this affects it 

> Comparison of int with float gives true if the values are numerically equal. Controversial, since floats are recommended to be compared using is_equal_approx.

this looks like a good thing, but I think it is game specific, most games won't need float precision or the is_equal_approx(), and if we do it like this it will only help in physics math & operations, but in normal comparing let's say I made a variable 4.0 for jumping, and I am using it to compare to something else that is integer, since this value isn't constantly changing no need to do the is_equal_approx(), so forcing it I think isn't good at all.

Note : I worked, and will continue working on a rollback netcode project which requires float precision (actually I don't use floats at all), but I get why is_equal_approx() feels important for you.

**Comment by LightningAA** [comment]
Also I can imagine frustrating bugs caused by users coming from JavaScript expecting stuff like `1 == "1"`, and it silently not working as expected. Maybe not, I mean there's already things that will surprise users coming from other languages, like integer division which works differently from python which uses `//`, I'm just speculating.

**Comment by Mickeon** [comment]
I think after JavaScript's constant and recommended usage of "**===**" it would be pretty safe to say that most programmers would really prefer to know if their comparison works as they imagine, without returning inconsistent or "unexpected" results as a result of the compiler attempting to lift some weight off their shoulders. The current error is a nudge towards that, I feel.

**Comment by Mickeon** [comment]
I suppose that implementing the function as it is written in the proposal would be better than changing the core `==`, but... what would be good use-cases, enough to include in the Global Scope? It's hard for me to imagine 'cause I really do believe it leads to unexpected behaviour in the long run.

**Comment by dalexeev** [comment]
> Also I can imagine frustrating bugs caused by users coming from JavaScript expecting stuff like `1 == "1"` to be true

I think it doesn't matter in this case. We cannot ensure that all expectations of all users are met. Here, convenience is more important, combined with predictability.

Comparison errors are not very convenient and most languages ​​don't have them (although the behavior differs from language to language).

Predictability is not the same as intuition/clairvoyance. Predictability is uniformity and the absence of exceptions. I suggest that the `==` operator in GDScript behave like `===` in some other languages. The alternative is adding `===` to GDScript, how do you like that suggestion?

> can compare a string with a number not on purpose which should give me an error

Everyone has a different perception. For me personally, the situation of random comparison of values ​​of different types and the possible errors associated with this seem unlikely. I make the most of static typing, but sometimes I need to compare the value of a dynamic type to a constant. I do this on purpose, and therefore such a restriction only annoys, not helps. I would prefer that even if I make a mistake, the comparison silently evaluates to false, the branch does not work, and I would debug this situation.

Also, now you may find that `match` is not equivalent to `if` - `elif` - `else`, since match compares the value exactly as I suggest. Agree, it's strange that the implicit comparison operator used in match is different from `==`.

```gdscript
if 1 == 1.0:
    print("a")
else:
    print("b")

# prints "a"

match 1:
    1.0:
        print("a")
    _:
        print("b")

# prints "b"

# You can this:
match x:
    "a": ...
    1: ...
    false: ...

# But you can not this:
if x == "a": ...
elif x == 1: ...
elif x == false: ...
# (There are situations when match is not suitable and you need to use if - elif - else.)
```

> I am also not sure when will ever anyone need to compare values of different type

Such cases happen. But not in the same way as in this proposal (`1 == "str"`), but for example, an element of an array/dictionary is compared with some constant. And not always a type mismatch is an unhandled error, often you just need to skip it. I will try to give examples of such situations. They don't happen very often, but they're annoying when they do happen.

**Comment by filipworksdev** [comment]
I disagree with the 1 == 1.0 should be false. 

There is no int or float in nature. This is just a limitation of computers representing real numbes. Both actually represent the exact same number which is real number 1 and should be true and equivalent.

Your proposed change would require everyone to type cast every single number comparison  which not only breaks all existing code but also will make coding much more difficult to write
`float(1) == 1.0`
`1 == int(1.0)`

The main feature proposed however is quite niche 1 == "str"

I would be in favor of === which does all the above as a much more strict type of == but keep == unchanged.

**Comment by dalexeev** [comment]
> disagree with the `1 == 1.0` should be false.
> 
> There is no int or float in nature.

> Your proposed change would require everyone to type cast every single number comparison which not only breaks all existing code but also will make coding much more difficult to write
> `float(1) == 1.0`
> `1 == int(1.0)`

In that case, you probably support #1866? I don't like the difference between `float` and `int` either, and would rather just have a `number`. But the reality is that there are too many differences between `int` and `float` in Godot/GDScript, and we only go in the direction of their distinction. For example, godotengine/godot#45303. Not only that, I see a lot of evidence for this trend, and I think I've written about it somewhere before.

The only "exception" is that you can pass an `int` where a `float` is expected, which is fine because it's convenient and doesn't break anything. But pretending that `2 == 2.0` is weird in my opinion, as there are many places where the two values ​​cause different behavior. `x / 2` and `x / 2.0` are not the same thing, `dict[2]` and `dict[2.0]` are not the same thing, `match` `2` and `2.0` are different, etc. This kind of convenience exceptions are wrong, in my opinion look (same as `null == freed_object` in 4.0-dev, see godotengine/godot#59816).

<details>
<summary>Meta offtopic</summary>

I love Godot, it has great architecture and a lot of brilliant ideas. But there are also a lot of basic things that don't go well with each other, simply because no one thought about it. I just pay attention to such points so that we think about them and maybe come up with something better, not necessarily exactly what I suggested. There are indeed problems, we just do not pay attention to them. I don’t see them right away either, I was going to write this proposal about 2 years ago, but I put it off. And during all this time my opinion has not changed, I found only confirmation in favor of this. Therefore, I ask you to also take some time to think before categorically denying this "simple" problem and limiting yourself to negative reactions (although I am also bad at this, but I usually prefer the emotion of confusion, not a thumbs down).

The Godot community is dominated by an atmosphere of pragmatism and idealism is scorned. But the truth is, there needs to be some healthy balance between these approaches.

Resistance to unreasonable changes is good. But it upsets me that when I point out some problem (no matter how significant it is, the main thing is that it exists), some far-fetched arguments are given in response. Show how the proposed change breaks or worsens something and compare it with the benefits it provides.

Perhaps this is just my subjective impression, but it seems to me that I gave several different arguments, and in response I received only an argument about possible unintentional errors and a suggestion from Calinou to use my function instead of a standard operator.

</details>

**Comment by dalexeev** [comment]
> I think after JavaScript's constant and recommended usage of "**===**" it would be pretty safe to say that most programmers would really prefer to know if their comparison works as they imagine

> I would be in favor of a more strict equality === which does all the above but keep == unchanged.

I already [suggested](godotengine/godot#31331) this before, but the community reacted negatively. And @KoBeWi referred to their [proposal](godotengine/godot#26375). This proposal is a resurrection of KoBeWi's proposal. Having two similar operators - is it bad or good?

**Comment by KoBeWi** [comment]
I don't really get the discussion here. It's intuitive that comparing variables of 2 different types yields `false`, because well, they can't be equal. It's rather normal in dynamic languages. I don't see how displaying an error is helpful in any way. If you expect `1`, it doesn't matter if you get `2` or `"two"` or `Vector2(2, 2)`; they are all a different value than you wanted.
I closed my proposal, because I didn't run into this error in years, so I just didn't need it anymore. I'm not against it though.

I don't use Javascript much, but I assume `==` and `===` exists, because of the abominations like `1 == "1" => true`. If different types are always false, `===` is useless.

Also I disagree with `1 != 1.0`, but I imagine the proposal can be implemented without this change.

**Comment by filipworksdev** [comment]
Hi guys. So I'm only against the `1 != 1.0` aspect of the proposal. Everything else is potentially useful.

Now that I think about it adding a new operator just for strict type comparison is a bit overkill and I highly doubt vnen would want to implement it. 

@KoBeWi's solution of implementing type comparison to `==`  minus the `1 != 1.0` is better.  

My only request is if this moves forward and you do get this change into Godot please add a GDScript warning too for example `Warning: Comparing values of different types` that would alert to the fact that you are doing a type comparison rather than a value comparison and that `==` is being a bit more forgiving in that regard.

**Comment by dalexeev** [comment]
> So I'm only against the `1 != 1.0` aspect of the proposal.

In my understanding, if `x == y` is true, then `f(x) == f(y)` must be true, where `f` is an [idempotent](https://en.wikipedia.org/wiki/Idempotence) function/action/operation. If `2 == 2.0` is `true`, then these things must also be true: `x / 2 == x / 2.0`, `dict[2] == dict[2.0]`, etc., because dividing and getting a value by a key (for an unchanged dictionary) are idempotent operations. And since this is not true, since we very often **distinguish** between `2` and `2.0`, it is obvious that these are **different**, **not equal** values.

This can be achieved in several ways: change `==` or add `===`. But in any case, it seems to me that such a thing should be standard, we should draw the attention of users to these details.

**Comment by wareya** [comment]
> In my understanding, if `x == y` is true, then `f(x) == f(y)` must be true, where `f` is an [idempotent](https://en.wikipedia.org/wiki/Idempotence) function/action/operation. If `2 == 2.0` is `true`, then these things must also be true: `x / 2 == x / 2.0`, `dict[2] == dict[2.0]`, etc., because dividing and getting a value by a key (for an unchanged dictionary) are idempotent operations.

I'm not sure what idempotency is doing here, but I that some of your assumptions are untrue. If your `f()` gets a value from a dictionary and that dictionary contains NaNs then `f(x) == f(x)` might very well be false. You can generate NaNs with normal math code by e.g. taking the log of a negative number:

```gdscript
    prints(-1, -1)
    prints(-1 == -1) # true
    prints(log(-1), log(-1))
    prints(log(-1) == log(-1)) # false
```

You can also create a NaN by creating an INF somehow (e.g. multiplying two very large numbers together, like 1.0e200) and then multiplying that INF by +/-0.0. You don't need special funky functions that have special NaN cases to generate NaNs; just multiplying by very big and very small numbers is enough.

```gdscript
    prints(1.0e200) # very big float
    prints(1.0e200 * 1.0e200) # INF
    prints(1.0e200 * 1.0e200 * 0.0) # NaN (compares falsely to itself)
```

Floating-point equality has never been a "same value" operator. If godot handled division by signed zero with signed infinities instead of errors (doing so would still be within the IEEE spec), you could demonstrate this without NaNs. `-0.0 == 0.0` would be true while `1.0/-0.0 == 1.0/0.0` would be false.

**Comment by dalexeev** [comment]
> ```gdscript
> prints(log(-1), log(-1))
> ```

The logarithm is only defined for positive numbers. `log(-1)` does not exist. `NaN` is strange in itself, it should not be understood as some kind of real value, because the real value cannot be unequal to itself. And in most cases, we work with normal values, not "crazy".

In a practical sense, the difference in types is more significant. **How does the fact that values of the same type can be unequal make it logical that values of different types can be equal?**

> `-0.0 == 0.0` would be true while `1.0/-0.0 == 1.0/0.0` would be false.

Mathematically -0 = +0 (except for limits in calculus). In terms of the binary representation of `float`, `-0.0` ≠ `+0.0`. But in essence, this is the same as Unicode normalization forms: strings that look the same can be binary different.

I mentioned above that it is recommended to compare `float`s with `is_equal_approx`, `is_zero_approx`, `is_nan`, `is_inf` rather than with `==`.

**Comment by dalexeev** [comment]
I added in a note about warnings in OP.

An alternative suggestion for those who are worried about possible unintentional bugs, especially when using dynamic typing. Turn the error into a warning and allow users to disable it in the Project Settings. By default, `1 == "str"` returns `false` and produces a warning. If I don't need the warning, then I just disable it.

**Comment by Mickeon** [comment]
> An alternative suggestion for those who are worried about possible unintentional bugs, especially when using dynamic typing. Turn the error into a warning and allow users to disable it in the Project Settings. By default, `1 == "str"` returns `false` and produces a warning. If I don't need the warning, then I just disable it.

Now this one I was thinking about would be nice. If the proposal was essentially worded like this (also excluded `1 == 1.0` because it's a WAY more debatable change of behaviour), I feel like the reception would've been more positive.

**Comment by filipworksdev** [comment]
> In my understanding, if `x == y` is true, then `f(x) == f(y)` must be true, where `f` is an [idempotent](https://en.wikipedia.org/wiki/Idempotence) function/action/operation. If `2 == 2.0` is `true`, then these things must also be true: `x / 2 == x / 2.0`, `dict[2] == dict[2.0]`, etc., because dividing and getting a value by a key (for an unchanged dictionary) are idempotent operations. And since this is not true, since we very often **distinguish** between `2` and `2.0`, it is obvious that these are **different**, **not equal** values.

This is not a mathematical problem but a computer science problem.  

In your example 2 and 2.0 are just different ways of representing the same mathematical value 2 from reality and therefore a mathematically they should be equivalent. I don't think nature has different values for number 2. Dig deeper, however, and you gonna start to find a whole slew of inconsistencies because they are really just very smart pragmatic solutions to the difficulty computers have in representing numbers.  So while they are equal doesn't mean they can be used interchangeably everywhere. For example you might find dict [0.0] is different than dict[-0.0]. Different ways of representing the same numbers from reality for different purpose. So while I find the float index enticing I would still prefer to use int 😁

**Comment by dalexeev** [comment]
> In your example 2 and 2.0 are just different ways of representing the same mathematical value 2 from reality and therefore a mathematically they should be equivalent.

`2` and `"2"` are also different representations of the same thing, but you are not suggesting that they be considered equal.

> This is not a mathematical problem but a computer science problem.

That is why I am suggesting this. And it makes sense specifically for GDScript. This doesn't make sense for JavaScript since it doesn't have `int` and `float`, only `Number`. This doesn't make sense in Python, since `int` and `float` are "the same" in most cases. For comparison:

1\.

```python
# python
print(5 / 2)   # 2.5
print(5 / 2.0) # 2.5
```

```gdscript
# gdscript
print(5 / 2)   # 2
print(5 / 2.0) # 2.5
```

2\.

```python
# python
d = {}
d[1] = "a"
d[1.0] = "b"
print(d) # {1: 'b'}
```

```gdscript
# gdscript
var d = {}
d[1] = "a"
d[1.0] = "b"
print(d) # {1:"a", 1:"b"}

print(var2str(d))
# {
# 1: "a",
# 1.0: "b"
# }
```

3\.

```python
# python
match 1:
    case 1.0:
        print("a")
    case _:
        print("b")
# a
```

```gdscript
# gdscript
match 1:
    1.0:
        print("a")
    _:
        print("b")
# b
```

Maybe if I start looking I'll find even more examples where GDScript treats `1` and `1.0` as different. And since they are different, they are not equal, everything is simple.

I'm not saying that GDScript should behave like Python. However, in the case of Python, it is logical why `1 == 1.0`, but in the case of GDScript it is not logical, not obvious, and not convenient from a practical point of view.

**Comment by YuriSizov** [comment]
> If you expect `1`, it doesn't matter if you get `2` or `"two"` or `Vector2(2, 2)`; they are all a different value than you wanted.

That's a dangerous way of thinking about it. If you expect `1`, but receive `"cucumber"` instead that doesn't just mean that it's not a value you want. It also means that somehow somewhere you've introduced an undefined behavior that lets values of a wrong type to funnel into your method that deals with numbers. If you're making a language parser, or some other generic input handler, sure, this can be a very normal thing to happen. But I don't think in scripting games this would be as normal. As games are cheap to crash, it's often safer to crash them when a UB is met. So the error makes sense to me.

Now, if you really don't care about having a UB in your project and would like to just throw the unexpected value away, I assume it's more important when that happens in release builds than in debug builds. So maybe the change here should only apply to release builds, to bring some stability to your released game when a UB happens, but keep it crashing/breaking in debug builds to let you actually know that something is terribly wrong with your code.

**Comment by dalexeev** [comment]
> That's a dangerous way of thinking about it. If you expect `1`, but receive `"cucumber"` instead that doesn't just mean that it's not a value you want.

If you expect `1`, but receive `1.0` instead that doesn't just mean that it's not a value you want. It also means that somehow somewhere you've introduced an undefined behavior that lets values of a wrong type to funnel into your method...

Why don't `int_val == float_val` and `str_val == null` throw an error in this case? We could require the user to explicitly cast `float(int_val) == int_val`. I am primarily for consistency and minimizing the number of exceptions.

See my comment above about replacing the error with a warning, I think this is the best compromise.

> This should be discussed in a proposal of its own.

I agree that this issue should be discussed separately. This proposal contains several important issues, and disagreement on one should not affect the others.

**Comment by YuriSizov** [comment]
> If you expect `1`, but receive `1.0` instead that doesn't just mean that it's not a value you want. It also means that somehow somewhere you've introduced an undefined behavior that lets values of a wrong type to funnel into your method...

You're being sarcastic, but I don't necessarily disagree that this is a potential problem. 🙃 Typing numbers in Godot and GDScript specifically is a hit and miss endeavor. You get all kinds of inconsistencies with logic, like rounding methods returning floats. Still, numbers are relative to each other, and in other languages and environments there are like a dozen other numeric types. I wouldn't equate receiving `1.0` in place of an int to receiving a `"cucumber"` in place of an int.

As for null comparisons, well, nulls are special, which is why they are both loved and hated. Still, wouldn't ever consider comparing a string to a null, so I as far as my sensibilities go, I think we could be rid of it.

**Comment by me2beats** [comment]
I am kinda for `===`
I had cases when I should compare arg1 and arg2, but I don't know if arg1, arg2 is a String or an Object.
So I had to create a function `are_equal(arg1, arg2)`.

That was a bit annoying.
Maybe I'm used to comparing objects of different types in Python, this is allowed there.

On the other hand I understand that these cases are not so usual, usually you want to compare values of one type. That's why I'm for adding new operator and leaving the old == as is

**Comment by dalexeev** [comment]
Closed by godotengine/godot#71758.
